[c10d] Add NCCL memory allocator

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #145675

This PR implements a small UI improvement over #133603.

It prepares a NCCL memory allocator in torch cpp and then pybind's it out, so that user can directly use it.

UI:
```
pool = torch.cuda.MemPool(backend.mem_allocator)
with torch.cuda.use_mem_pool(pool):
    tensor = torch.arange(1024 * 1024 * 2, device=device)
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[c10d] Add NCCL memory allocator

[ghstack-poisoned]

wconstab commented:
the API looks nice. I have 2 high level questions about the direction

1) do we have any input from any other vendors who want to support similar functionality?  It would be nice if the allocator creation and backend-registration were members of 'backend' object accessed through processgroup API so that any backend can easily override the behaviors cc @c-p-i-o @xw285cornell @sujoysaraswati

2) it seems possible to either use NVLS automatically (via compile/inductor) or manually by calling this API.  Is this API also useful for inductor binaries?  (I don't see why not..) cc @yifuwang

syed-ahmed reviewed (approved):
LGTM!

kwen2501 committed:
Update on "[c10d] Add NCCL memory allocator"


This PR implements a small UI improvement over #133603.

It prepares a NCCL memory allocator in torch cpp and then pybind's it out, so that user can directly use it.

UI:
```
pool = torch.cuda.MemPool(dist.nccl_mem_allocator)
with torch.cuda.use_mem_pool(pool):
    tensor = torch.arange(1024 * 1024 * 2, device=device)
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab reviewed (approved):
Thanks for the changes!  This seems good to me.  It might be enough just as it is.  If we eventually want to make a smoother API for getting an allocator, i think we could build that strictly in the python layer later on and use these apis under the hood.

kwen2501 commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "This fails to build internally, can you please take a look at D68831004 for more details?"

kwen2501 committed:
Update on "[c10d] Add NCCL memory allocator"


This PR implements a small UI improvement over #133603.

It prepares a NCCL memory allocator in torch cpp and then pybind's it out, so that user can directly use it.

UI:
```
pool = torch.cuda.MemPool(backend.mem_allocator)
with torch.cuda.use_mem_pool(pool):
    tensor = torch.arange(1024 * 1024 * 2, device=device)
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
Added `#ifdef` guards. Re-merging.
@pytorchbot merge

kwen2501 committed:
Update on "[c10d] Add NCCL memory allocator"


This PR implements a small UI improvement over #133603.

It prepares a NCCL memory allocator in torch cpp and then pybind's it out, so that user can directly use it.

UI:
```
pool = torch.cuda.MemPool(backend.mem_allocator)
with torch.cuda.use_mem_pool(pool):
    tensor = torch.arange(1024 * 1024 * 2, device=device)
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@ZainRizvi Arg, sorry I forgot to add an `#else`, too stupid. 
Fixed now and trying to reland again. Sorry.
